text/000-buildpacks-contribute-default-process-type.md

```diff
@@ -0,0 +1,180 @@
+
+Existing buildpack users would need to be taught about the change so that (1) they no longer have to pass a flag to specify the default process type or (2) they now need to pass a flag to explicitly override the default that buildpacks choose. For (2), it's likely that the number of users in this category would be small given that the lifecycle (pre- platform API 0.4) and pack (post- platform API 0.4) both encourage `web` as the default, and it's likely that buildpacks would also encourage this same default.
+
+Buildpack authors would need to be taught how to configure the default process type.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+The changes from this RFC should require a bump to the buildpack API.
+
+Currently, during the `build` phase, buildpacks may declare process types by writing entries to `<layers>/launch.toml`. During the `export` phase, the `<layers>/launch.toml` from each buildpack are aggregated to produce a combined processes list "such that process types from later buildpacks override identical process types from earlier buildpacks" ([spec](https://github.com/buildpacks/spec/blob/main/buildpack.md#process-3)).
+
+This RFC proposes a new `default` key in `<layers>/launch.toml`:
+
+```
+[[processes]]
+type = "<process type>"
+command = "<command to run>"
+default = <boolean (default=false)>
+```
+
+Example:
+
+```toml
+[[processes]]
+type = "web"
+command = "bundle exec ruby app.rb"
+default = true
+
+[[processes]]
+type = "worker"
+command = "bundle exec ruby worker.rb"
+```
+
+Buildpacks may set `default = true` to indicate that the process type being defined should be the default process type for the app image. If not specified, no default designation will be assumed (to be consistent with buildpacks implementing earlier buildpack APIs). If multiple buildpacks try to set the default process type, the latest buildpack will win.
+
+If a buildpack attempts to define two processes with `default = true` specified, the lifecycle will fail.
+
+#### `pack` migration plan
+
+`pack build` has a `--default-process` flag. If this flag is not provided by the user, `pack` will provide `-process-type web` to the lifecycle. As currently architected, this will prevent any default process type contributed by buildpacks from ever taking hold, as the lifecycle will prefer the provided `-process-type` to any default process type contributed by buildpacks.
+
+Migration plan:
+
+`pack` stops sending `-process-type web` when `CNB_PLATFORM_API` is greater than `0.X`. The lifecycle will assume `default = true` for `web` processes from buildpacks implementing buildpack APIs earlier than `0.Y`. The lifecycle could print an info message when buildpacks do not specify a default process type. `-process-type <unknown process type>` could then become a failure case (vs. warn, what it is today).
```

Comment:
Do we define `0.X` and `0.Y` anywhere?

Reply:
Nope. I'm not sure that they need to be any particular value for the RFC to function as intended, but practically speaking it seems like they would both be 0.6.

Reply:
well, from discussions I assume you mean by `0.X` you mean "a version of the Platform API that does not include this RFC" and `0.Y` is a version of the Buildlpack API that doesn't include this RFC.